pdf-converter/client.py

```diff
@@ -0,0 +1,427 @@
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+import asyncio
+import logging
+import os
+import subprocess
+import sys
+from collections import namedtuple
+from pathlib import Path
+from PIL import Image
+from tempfile import TemporaryDirectory
+
+PROG_NAME = Path(sys.argv[0]).name
+ARCHIVE_PATH = Path(Path.home(), "QubesUntrustedPDFs")
+
+MAX_PAGES = 10000
+MAX_IMG_WIDTH = 10000
+MAX_IMG_HEIGHT = 10000
+DEPTH = 8
+
+Dimensions = namedtuple("Dimensions", ["width", "height", "size", "depth"])
+Representation = namedtuple("Representations", ["initial", "final"])
+
+
+class DimensionError(ValueError):
+    """
+    """
+
+
+class PageError(ValueError):
+    """
+    """
+
+
+class ReceiveError(Exception):
+    """
+    """
+
+
+class RepresentationError(ValueError):
+    """
+    """
+
+
+###############################
+#         Utilities
+###############################
+
+
+def check_paths(paths):
+    abs_paths = []
+
+    for path in [Path(path) for path in paths]:
+        abspath = Path.resolve(path)
+
+        if not abspath.exists():
+            logging.error(f"No such file: \"{path}\"")
+            sys.exit(1)
+        elif not abspath.is_file():
+            logging.error(f"Not a regular file: \"{path}\"")
+            sys.exit(1)
+
+        abs_paths.append(abspath)
+
+    return abs_paths
+
+
+def check_range(val, upper):
+    if not 1 <= val <= upper:
+        raise ValueError
+
+
+def unlink(path):
+    """Wrapper for Path.unlink(path, missing_ok=True)"""
+    try:
+        path.unlink()
+    except FileNotFoundError:
+        pass
+
+
+async def cancel_task(task):
+    if not task.done():
+        task.cancel()
+        await task
+
+
+async def wait_proc(proc):
+    await proc.wait()
+    if proc.returncode:
+        raise subprocess.CalledProcessError
+
+
+async def terminate_proc(proc):
+    if proc.returncode is None:
+        proc.terminate()
+        await proc.wait()
+
+
+###############################
+#       Qrexec-related
+###############################
+
+
+async def recv_b(proc, size):
+    """Qrexec wrapper for receiving binary data from the server"""
+    try:
+        untrusted_data = await proc.stdout.readexactly(size)
+    except asyncio.IncompleteReadError:
+        logging.error("server may have died...")
+        raise
+
+    if not untrusted_data:
+        raise EOFError
+
+    return untrusted_data
+
+
+# TODO (?): Size limit for readline()
+async def recvline_b(proc):
+    """Qrexec wrapper for receiving a line of binary data from the server"""
+    untrusted_data = await proc.stdout.readline()
+
+    if not untrusted_data:
+        raise EOFError
+
+    return untrusted_data
+
+
+async def recv(proc, size):
+    """Convenience wrapper for receiving text data from the server"""
+    try:
+        untrusted_data = (await recv_b(proc, size)).decode()
```

Comment:
`recv()` actually is never used by the client so I took it out earlier today. Besides, `recvline()` is the one used for image dimensions so an `ascii` encoding would work there. But to limit general text data to `ascii` is a bit much I think. Not that it matters anyways though since we never used the function.